### ISSUE TITLE
FindInFilesDomain uses fs.readFileSync

### ISSUE THREAD

**Issue by zaggino** [adobe/brackets#11613]
Hi guys, why do we use sync file read operation in the domain? https://github.com/adobe/brackets/blob/master/src/search/node/FindInFilesDomain.js#L200

Also, we should be checking if a file is not binary, with something like this: https://www.npmjs.com/package/isbinaryfile

This is blocking other domain files which often fail to load on Brackets' startup. Related issue https://github.com/zaggino/brackets-git/issues/1149

cc authors @abose @prksingh @rroshan1 @prafulVaishnav

**Comment by abose** [comment]
There are usually large number of files in the project; when we queue large number of async file reads in the queue- it will result in node being unresponsive until all the files got indexed as the event loop was filled with file io related events. To circumvent this, we used synchronous file reads, as this will ensure that only one file op is pending at a time and each file read is interleaved with a `setimmediate` call so that the next file read event happens only after any pending events in queue is processed. It was mostly effective, we used to get frequent node failures immediately with the async file reads. But it is not as widely effective as we wanted it to be. 

What we can do is, 
1. right now 10 files are read as 1 synchronous operation- we can shorten this to just 1.
2. improving on 1 if required, we can set a timeout of 10(0) MilliSecs between each interleaves to yield to other events.
3. Move indexing to a different node process fork - This was our original implementation, but having to manage 2 node processes and sending/synchronising events across 2 process boundaries[brackets-node1-indexnode] was making it difficult to implement. It was also slower for instant search due to the data transfer required across 4 process boundaries.

If 1&2 fixes it would be the simplese fix and just leaves one node process.
 3 is a much more difficult route to take, but as we have stabilised search, the things left would be process management and data transfer optimisations, which would be manageable to do. 

Any other idea on optimizing this is welcome :)

**Comment by abose** [comment]
The binary file check would be straight forward to implement.

**Comment by zaggino** [comment]
Id suggest doing a queue of async reads, so next one starts when previous finishes. This way its non blocking and similarly fast to pure sync approach.

Plus Id put a binary check in place and maybe dont process files over a certain size, 10mb maybe?

**Comment by abose** [comment]
The size restriction we can enforce, and a preference to override it. Also the bin file detection should be made more robust. The async change will not solve the large file read issue. I will try chaining the async file reads if 1&2 fails to deliver. Our initial impl was async, but it was 4-5 times slower than the sync methods, so instead of 10 secs of node heavy loads at startup, we were seeing up to 50 secs of heavy loads and that gave a very bad exp. I'll check out some optimizations with the current impl and see what it does.

**Comment by zaggino** [comment]
Ok, let me know if you'd want any help with this.

**Comment by zaggino** [comment]
Other thought ... does the indexing start right after you start Brackets? Would it be possible to delay it for a few seconds? (ideally wait for an event when all extensions are loaded, but that'd be hard to figure out, maybe hard-coded 5 seconds would be enough).

This way, the extensions won't be blocked when trying to init by indexing and it wouldn't concern user much, very little people start they editor just to do immediate Find in Files search.

**Comment by abose** [comment]
We already have a 5 second timeout on app init before an instant search kicks in. i think i will increse it to 20 seconds?

**Comment by zaggino** [comment]
ah, didn't knew that maybe to 10 at least ... together with binary file check it should be enough not to block other node stuff

**Comment by petetnt** [comment]
Is this landing on 1.5? I think it is causing high unstability for @zaggino's brackets-eslint extension for: for some reason I cannot even get it stay up once without running to `ESLint terminated with error: Max connection attempts reached` or ESLint getting terminated because it cannot find `eslint-react-plugin` in time.
`

**Comment by zaggino** [comment]
@abose Another issue with this feature:

I have my `node_modules` excluded from searching (they contain a ton of files...) but they still get indexed by the node_domain -> I need to search only 397 files, not 19646.

![image](https://cloud.githubusercontent.com/assets/1067319/9709786/27120d26-5575-11e5-821a-5477a8ff814c.png)

**Comment by abose** [comment]
That is strange, i had it index only the filtered file list as i remember. Will check this.
I was on vacation :)- I will raise a pr soon with the fixes.

**Comment by zaggino** [comment]
I've done this alteration in my domain (because I had problems loading some of my projects):
```
try {
            var stats = fs.statSync(filePath);
            if (stats.size > 1000000) {
                console.warn('[warn] File too large for sync file search (' + stats.size + '): ' + filePath);
            }
            projectCache[filePath] = fs.readFileSync(filePath, 'utf8');
        } catch (ex) {
            console.log(ex);
            projectCache[filePath] = null;
        }
```
and can clearly see that most of the >1MB files come from `node_modules` (builded and/or minified versions of modules) which leads me to believe that the exclusion set is not being used in this case.

**Comment by zaggino** [comment]
@petetnt for quick solution, using https://github.com/zaggino/brackets-file-tree-exclude#how-to-install solves the problem, I use it to exclude `node_modules` from Brackets and `brackets-eslint` never fails to load. The issue that it can't find `eslint-react-plugin` might be an extension issue.

**Comment by petetnt** [comment]
Thanks for the @zaggino, however I am already excluding files (`node_modules`, `.git`, `bower_components`, `dist`-folders and other usual suspects using gruehle/exclude-folders). I agree that the issue is that it can't find the globally installed `eslint-react-plugin`, but I think it's just that Node timing out while trying to find it. (I am often working on files on a network folder, which might make the indexing even slower...)

**Comment by zaggino** [comment]
My performance issues that lead to opening this issue were addressed in #11675 so closing this. If anyone still has problems with latest master, we can open another issue or reopen this one.

**Comment by zaggino** [comment]
@petetnt you can fully disable indexing when working with network folder through option `"findInFiles.nodeSearch": false,`
